### ISSUE TITLE
Trying to do Tracing + Scripting with PYTORCH_JIT=0 silently breaks things

### ISSUE THREAD

**Issue by ShadowLau** [pytorch/pytorch#18324]
## ❓ Questions and Help

Hi, recently i am trying to use pytorch model in C++，i mixed tracing and scriping to transform pytorch model. But i find the outputs in C++ is different with outputs in python, and the outputs is different w/o PYTORCH_JIT=0 in pure python. I load pretrained model before tracing(maybe i was wrong with that, but how to load weights before tracing if the traced modules are submodules of a ScriptModule?).

Is there someone who ever encounterd this problem?

Below is my scripts.

```
import torch
import torch.nn as nn


class SubNet2(torch.jit.ScriptModule):
    __constants__ = ['pool']

    def __init__(self):
        super(SubNet2, self).__init__()
        self.backbone_fea_dim = 2048
        self.fea_dim = 256
        bottleneck_size = [256, 128]
        print("in 1 ...")
        reduce_dim_z = nn.Sequential(
                              nn.Conv2d(self.backbone_fea_dim * 2, self.fea_dim, bias=False, kernel_size=1, padding=0),
                              nn.BatchNorm2d(self.fea_dim, eps=0.001),
                              nn.ReLU())
        reduce_dim_z.load_state_dict(torch.load('./models/weights/reduce_dim_dict2.pkl'))
        self.reduce_dim_z = torch.jit.trace(reduce_dim_z, torch.rand(1, self.backbone_fea_dim * 2, 8, 8))
        
        print("in 2 ...")
        s_atten_z = nn.Sequential(
            nn.Conv2d(1, bottleneck_size[1], kernel_size=8, padding=0, bias=False),
            nn.ReLU(inplace=True),
            nn.Conv2d(bottleneck_size[1], 64, kernel_size=1, padding=0, bias=False))
        s_atten_z.load_state_dict(torch.load('./models/weights/sz_dict2.pkl'))
        self.s_atten_z = torch.jit.trace(s_atten_z, torch.rand(1, 1, 8, 8))
        
        print("in 3 ...")
        c_atten_z = nn.Sequential(
            nn.AvgPool2d(8),
            nn.Conv2d(bottleneck_size[0], self.backbone_fea_dim, kernel_size=1, padding=0, bias=False))
        c_atten_z.load_state_dict(torch.load('./models/weights/cz_dict2.pkl'))
        self.c_atten_z = torch.jit.trace(c_atten_z, torch.rand(1, 256, 8, 8))
        
        print("in 4 ...")
        t_info = nn.Sequential(
            nn.Conv3d(self.backbone_fea_dim, 256, 1, bias=False),
            nn.BatchNorm3d(256),
            nn.ReLU(inplace=True),
            nn.Conv3d(256, self.fea_dim, 3, padding=1, bias=False),
            nn.BatchNorm3d(self.fea_dim),
            nn.ReLU(inplace=True))
        t_info.load_state_dict(torch.load('./models/weights/tf_dict2.pkl'))
        self.t_info = torch.jit.trace(t_info, torch.rand(1, self.backbone_fea_dim, 10, 8, 8))
        
        print("in 5 ...")
        self.pool = nn.Sequential(nn.AvgPool2d(8))

    @torch.jit.script_method
    def generate_attention_z(self, x):
        z = self.reduce_dim_z(x)
        atten_s = self.s_atten_z(z.mean(dim=1, keepdim=True)).view(z.size(0), 1, z.size(2), z.size(3))
        atten_c = self.c_atten_z(z)
        z = torch.sigmoid(atten_s * atten_c)
        return z, 1 - z

    @torch.jit.script_method
    def forward(self, x):
        if x.dim() == 4:
            x = x.view((1,) + x.size())

        video_num = x.size(0)
        depth = x.size(1)

        res = torch.cat((x[:, 0].contiguous().view((x.size(0), 1) + x.size()[2:]), x), dim=1)
        res = res[:, :-1]
        res = x - res
        h = x[:, 0]
        output = []
        for t in range(depth):
            con_fea = torch.cat((h - x[:, t], res[:, t]), dim=1)
            z_p, z_r = self.generate_attention_z(con_fea)
            h = z_r * h + z_p * x[:, t]
            output.append(h)

        fea_t = torch.stack(output, dim=2)
        fea_t = self.pool(self.t_info(fea_t).mean(dim=2)).view(video_num, -1)
        return fea_t


if __name__ == '__main__':
    net = SubNet2()
    net.eval()
    input_data = torch.ones(2, 2048, 8, 8)
    output = net(input_data)
    print(output.shape)
    print(output)


```

[The model link](https://drive.google.com/file/d/1fewpw_pGkqxvL4hq3GdwMelnpL7Y1LB8/view?usp=sharing)

**Comment by t-vi** [comment]
I will say that it's pathological to feed a constant to BatchNorm, but nonetheless something bad seems to happen with `PYTORCH_JIT=0` and the BatchNorm.

```python
import torch
import torch.nn as nn


class SubNet2(torch.jit.ScriptModule):
    def __init__(self):
        super().__init__()
        t_info = nn.Sequential(
            nn.BatchNorm2d(2),
        )
        self.t_info = torch.jit.trace(t_info, torch.rand(1, 2, 8, 8))

    @torch.jit.script_method
    def forward(self, fea_t):
        i = self.t_info(fea_t)
        return i

if __name__ == '__main__':
    net = SubNet2()
    net.eval()
    input_data = torch.ones(2, 2, 8, 8)
    output = net(input_data)
    print ("mean", output.mean().item(), "std", output.std().item())


    t_info = nn.Sequential(
        nn.BatchNorm2d(2),
    )
    output = t_info(input_data)
    print ("mean", output.mean().item(), "std", output.std().item())
```

**Comment by ShadowLau** [comment]
@t-vi Thanks for reply, but in your scripts, if the "input_data" is generated by torch.rand(), the outputs are different.

**Comment by ShadowLau** [comment]
@t-vi Thanks

**Comment by t-vi** [comment]
I'm not sure it should silently break though. @driazati @jamesr66a , do you know?

**Comment by SsnL** [comment]
Trace with PYTORCH_JIT=0 does a lot of things wrong. Relevant issue: https://github.com/pytorch/pytorch/issues/16293

**Comment by t-vi** [comment]
Thanks, Simon. I'm closing this as a duplicate of the bug with your much more elaborate analysis.
